cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -1384,9 +1385,342 @@ size_t CacheAllocator<CacheTrait>::wakeUpWaitersLocked(folly::StringPiece key,
+      break;
+    case dml::status_code::bad_length:
+      /* Invalid number of elements was specified */
+      error = "bad len";
+      break;
+    case dml::status_code::inconsistent_size:
+      /* Input data sizes are different */
+      error = "inconsis size";
+      break;
+    case dml::status_code::dualcast_bad_padding:
+      /* Bits 11:0 of the two destination addresses are not the same */
+      error = "dualcast bad padding";
+      break;
+    case dml::status_code::bad_alignment:
+      /* One of data pointers has invalid alignment */
+      error = "bad align";
+      break;
+    case dml::status_code::buffers_overlapping:
+      /* Buffers overlap with each other */
+      error = "buf overlap";
+      break;
+    case dml::status_code::delta_bad_size:
+      /* Invalid delta record size was specified */
+      error = "delta bad size";
+      break;
+    case dml::status_code::delta_delta_empty:
+      /* Delta record is empty */
+      error = "delta emptry";
+      break;
+    case dml::status_code::batch_overflow:
+      /* Batch is full */
+      error = "batch overflow";
+      break;
+    case dml::status_code::execution_failed:
+      /* Unknown execution error */
+      error = "unknw exec";
+      break;
+    case dml::status_code::unsupported_operation:
+      /* Unknown execution error */
+      error = "unsupported op";
+      break;
+    case dml::status_code::queue_busy:
+      /* Enqueue failed to one or several queues */
+      error = "busy";
+      break;
+    case dml::status_code::error:
+      /* Internal library error occurred */
+      error = "internal";
+      break;
+    case dml::status_code::ok:
+      /* should not be here */
+      error = "ok";
+      break;
+    default:
+      error = "none of the above - okay!!";
+      break;
+  }
+  return error + " mem copy error";
+}
+
+using allocator_t = std::allocator<dml::byte_t>;
+using batch_handler_t = dml::handler<dml::batch_operation, allocator_t>;
+
+template <typename CacheTrait>
+void CacheAllocator<CacheTrait>::evictRegularItems(TierId tid, PoolId pid, ClassId cid,
+                                                   std::vector<EvictionData>& evictionData,
+                                                   std::vector<WriteHandle>& newItemHdls,
+                                                   bool skipAddInMMContainer,
+                                                   bool fromBgThread,
+                                                   std::vector<bool>& moved) {
+  /* Split batch for DSA-based move */
+  const auto& pool = allocator_[tid]->getPool(pid);
+  const auto& allocSizes = pool.getAllocSizes();
+  auto isLarge = allocSizes[cid] >= config_.largeItemMinSize;
+  auto dmlBatchRatio = isLarge ? config_.largeItemBatchEvictDsaUsageFraction :
+                                 config_.smallItemBatchEvictDsaUsageFraction;
+  size_t dmlBatchSize =
+      (config_.dsaEnabled && evictionData.size() >= config_.minBatchSizeForDsaUsage) ?
```

Comment:
If there is no DSA device - we use regular DML software path?

Reply:
The implicit assumption is if dsaEnabled flag is set to true, there is at least one DSA device enabled.

handler = dml::submit<dml::hardware>(dml::batch, sequence); 

In the code currently, if dsaEnabled flag is wrongly set to true without enabling any DSA device, it would trigger a runtime exception. I can set it to dml::automatic which would take care of this scenario but we would lose explicit control over whether a request goes to software or hardware.